Using the material Grid component in Calypso

scruffian opened the pull request:
#### Changes proposed in this Pull Request

* This tries to implement the Material grid in Calypso: https://material-ui.com/layout/grid/
* I have implemented the grid in http://calypso.localhost:3000/stats/day/ben.blog

Before: 
<img width="1440" alt="Screenshot 2019-05-08 at 17 06 40" src="https://user-images.githubusercontent.com/275961/57390336-0137d900-71b4-11e9-85b7-f8d490a559ba.png">
<img width="1262" alt="Screenshot 2019-05-08 at 17 07 16" src="https://user-images.githubusercontent.com/275961/57390332-ff6e1580-71b3-11e9-932e-36f6ddfc0eab.png">
<img width="880" alt="Screenshot 2019-05-08 at 17 07 38" src="https://user-images.githubusercontent.com/275961/57390329-ff6e1580-71b3-11e9-89ab-b9cb209bd5ec.png">

After:
<img width="1440" alt="Screenshot 2019-05-08 at 17 06 51" src="https://user-images.githubusercontent.com/275961/57390335-009f4280-71b4-11e9-9558-8b5a8dad5c10.png">
<img width="1263" alt="Screenshot 2019-05-08 at 17 07 07" src="https://user-images.githubusercontent.com/275961/57390333-0006ac00-71b4-11e9-840c-a6fed97b1f48.png">
<img width="880" alt="Screenshot 2019-05-08 at 17 07 46" src="https://user-images.githubusercontent.com/275961/57390327-fed57f00-71b3-11e9-82f6-d72bb7f691d4.png">

The major changes are:
- The stacking order of the columns is different - this isn't easy to achieve using the Grid component, and I'm not convinced its worth extra code to achieve it.
- The spacing between the columns is very slightly different.
- The breakpoints are slightly different.

This seems like a simple way to move forward with adding a Grid to Calypso, however we don't seem to be able to specify flexible gutters.

scruffian committed:
Using the material Grid component in Calypso

jffng committed:
Change sm prop to xs on grid items
Adjust margin / padding to achieve consistent vertical spacing

jffng commented:
Made some minor changes to get the spacing right. 

> The stacking order of the columns is different - this isn't easy to achieve using the Grid component, and I'm not convinced its worth extra code to achieve it.

This doesn't resolve the stacking order, but adding `justify="flex-end"` to the container solves this issue at the `md` breakpoint:

<img width="837" alt="Screen Shot 2019-05-08 at 1 47 44 PM" src="https://user-images.githubusercontent.com/5375500/57399394-1d7d4b00-719f-11e9-87af-ced60c670650.png">

> ... we don't seem to be able to specify flexible gutters.

Should we move the grid component into Calypso and alter it to our needs? The non-flexible gutters plus different breakpoints seem like good reasons to do so.

jffng reviewed (commented):
Made some minor changes to get the spacing right. 

> The stacking order of the columns is different - this isn't easy to achieve using the Grid component, and I'm not convinced its worth extra code to achieve it.

This doesn't resolve the stacking order, but adding `justify="flex-end"` to the container solves this issue at the `md` breakpoint:

<img width="837" alt="Screen Shot 2019-05-08 at 1 47 44 PM" src="https://user-images.githubusercontent.com/5375500/57399394-1d7d4b00-719f-11e9-87af-ced60c670650.png">

> ... we don't seem to be able to specify flexible gutters.

Should we move the grid component into Calypso and alter it to our needs? The non-flexible gutters plus different breakpoints seem like good reasons to do so.

scruffian commented:
I opened an issue for the flexible gutters thing: https://github.com/mui-org/material-ui/issues/15631

I'm not sure about the breakpoints thing. It might make sense to move to Material breakpoints anyway, in which case this would work fine. As the <Grid> component was implemented we'd gradually migrate towards the new breakpoints...

jffng commented:
SGTM. We will follow up on breakpoints and responsive gutters, and ship iteratively.

jffng commented:
Given #32869, @blowery @griffbrad do you have thoughts on using the Material grid component directly in Calypso?

blowery commented:
@jffng Using `@material-ui/*` just to get a grid seems like a big leap. 19k of zipped JS with it's own style solution is a lot to bite off. We've looked at using various third-party material solutions before and have always ended up unhappy with the trade-offs entailed vs. using core components or our existing components. They also seem to be headed the way of CSS-in-JS in the next major, which will further complicate the integration story and bind us to their tech choices.

I'd be happy to fork it and have us work up an alternative inside the package in #32869 that uses our styling solution. Are we happy with the API it provides? What would we change if we owned it?

jffng commented:
Thanks @blowery!

> I'd be happy to fork it and have us work up an alternative inside the package in #32869 that uses our styling solution.

👍

> Are we happy with the API it provides? What would we change if we owned it?

Not exactly. It lacks an ability to create flexible gutters at different viewports. In addition, we'd alter the internals of the breakpoints.

jffng committed:
Adapt material grid into calypso-ui style component
Remove @material-ui/core dependency

jffng commented:
@scruffian summary of changes in the latest commit: 
- Pulled the material grid component and translated their js-generated styles into component scss
- Set the standard way the grid should behave directly in the styles (i.e. you don't need to set component props to get a 4, 8, and 12 col grid at xs, md, and lg viewports. Same goes for gutter spacing).
- Adjusted the breakpoints to Calypso
- Added support for adjusting the spacing via props

Figured we can test it here while `calypso-ui` package is in draft.

scruffian commented on client/components/grid/style.scss:
```diff
@@ -0,0 +1,203 @@
+    }
+}
+
+@media (min-width:660px) {
```
If we're going to use Calypso breakpoints then let's also use the Calypso breakpoint mixin.

scruffian commented:
This looks great. Might be a good idea to add a README so people know how to use it.

> Added support for adjusting the spacing via props

Do we want to allow this kind of flexibility?

jffng committed:
Bake in the material spacing rules to the component css

jffng commented:
Latest commit simplifies the component props and css by removing support for adjusting the spacing. Thoughts on this approach? I like its simplicity. 

Also going to add a README.

scruffian commented:
> Set the standard way the grid should behave directly in the styles (i.e. you don't need to set component props to get a 4, 8, and 12 col grid at xs, md, and lg viewports. Same goes for gutter spacing).

Does this mean you can remove some props in the Stats example?

jffng committed:
Adding docs
Aligning grid component behavior and API to the official material component adapted to calypso breakpoints

jffng committed:
Update stats example to work with latest grid component api and default props

jffng commented:
Took another pass at aligning this component to our needs. I believe the props are the simplest they can be for desired flexibility. 

Can you have a look at the README and let me know if it makes sense?

jffng committed:
Fix table formatting

scruffian commented:
Looks good. A few more thoughts:

1. Would it make sense to move this to `@wordpress/components` so it can be shared wider than just Calypso?

2. Is the `xs`, `sm`, `md`, `lg`, `xl` abstraction useful? In our breakpoints we reference the sizes directly. I wonder if we could do that here, so that the intention is more explicit?

jffng commented:
1. Good question. @griffbrad do you have thoughts on moving this to `@wordpress/components`? 

2. How would the props look to specify the number of columns a grid item should span?

scruffian commented:
> How would the props look to specify the number of columns a grid item should span?

This is a tricky one. I guess we can't use props like `>660`? Maybe `gt660`?

jffng commented:
Something like `gt660` maps to the existing Calypso breakpoint mixin, but the abbreviation + number don't feel much clearer than having to refer to a size in the README. 

The official [material web component](https://github.com/material-components/material-components-web/tree/master/packages/mdc-layout-grid) uses `phone`, `tablet`, and `desktop` which reads nicely, but may be limiting to have only three sizes. Perhaps it's a good place to start though?

Can you think of anyone else whose input would be valuable? Are there meaningful blockers to shipping this and iterating?

scruffian commented:
It was @MichaelArestad's idea to stick to numbers in the original breakpoints mixin, so I'd be interested to get his input here.

The main blocker to shipping is that if we decide to change the implementation it's going to be more work to update things...

jffng committed:
Fixing the flexbox classnames

jffng committed:
Fix the column stacking

jffng committed:
Merge branch 'try/material-grid' of github.com:Automattic/wp-calypso into try/material-grid

MichaelArestad commented:
We tried naming with more general names, but the numbers were clearer and more flexible down the road. Abstracting sizes into words is difficult. We didn't know how many we needed or how flexible it would need to be. At this point, I think we won't need to expect more than five sizes. `xs, sm, md, lg, xl` seem to fit the bill nicely.

> How would the props look to specify the number of columns a grid item should span?

I thought the `xs, sm, md, lg, xl` abstractions were designed for just that. From the readme, it seems each takes a value to show the number of columns an item should span at that size.

MichaelArestad commented:
After working with this implementation, it's become clear this is a bigger challenge than at first glance.

I rewrote a good portion of `stats/site.jsx` and some of the surrounding Sass files in order to fully utilize the grid on something like this. I ended up with this:

![image](https://user-images.githubusercontent.com/1123119/57887052-d5d47000-77eb-11e9-926c-5f926006cb98.png)

* Each stats module has been wrapped with a Grid item
* The surrounding CSS was written to allow the containers to wrap properly in the right order instead of using the columns
* I made it so each module will expand vertically to fill the row based on the tallest item in the row

## Problems encountered
Now, I wouldn't judge this aesthetically too much and the code does break other areas. However, I found some difficulties:

### Too many children
Having to reapply all sorts of flex properties to the children of a Grid item in order to take advantage of things like `align-content: stretch` is quite difficult without familiarity with Flexbox. I found myself wanting the abilities of the Grid component to be baked in to some of the more common containers like Card. Stats is an older section of Calypso so its components aren't written as nicely as newer ones. There are some classless divs that wrap many of the components.

### No concept of a subgrid
I found I needed to apply `display: flex` to each Grid item. The way I approached it was to add an additional class to target for just those items. Perhaps this isn't super necessary, but could be handled with a prop. It could also be useful if we add in CSS Grid support since subgrid is an actual thing there.

### CSS Grid exists
Now that CSS Grid is out, this component name is misleading. I wonder how we can either make the component flexible enough to do _either_ flexbox or CSS grid depending on the need. There are many parallels.

Other than those issues, the component itself seems thoughtfully-designed.

jffng commented:
Thanks for trying it out @MichaelArestad! 

To make sure I understand, is "too many children" and "having to reapply all sorts of flex properties" actually the same problem, or are there two different ones there? I think it's a good point you're raising and it's maybe one that a lot of UI libraries encounter: the need for custom rules when assembling complex components and layouts: https://medium.com/material-ui/introducing-material-ui-design-system-93e921beb8df   

> CSS Grid exists

What would you want the ability to do with CSS Grid that isn't possible with this component?

MichaelArestad commented:
> What would you want the ability to do with CSS Grid that isn't possible with this component?

@jffng [This article](https://medium.com/youstart-labs/beginners-guide-to-choose-between-css-grid-and-flexbox-783005dd2412) does a pretty good job of explaining the power of each. The gist is that Flexbox excels at one dimensional content (either a single row or a single column) while Grid excels at two dimentional content (rows and columns AKA grids).

They are both useful for different purposes. We tend to use Flexbox to tackle what Grid can do better (like the Stats section) because it's a little more familiar and with our skills, we can probably make anything do anything. Heck, we've all written our own versions of flexbox before flexbox was ever around (minus the whole non-hacky vertical centering thing and even then...).

Does that make sense?

sixhours commented:
Given we're retiring Muriel, I'm going to close this out for now.
